bin/bout-v5-header-upgrader.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python3
+
+import argparse
+import copy
+import difflib
+import re
+import textwrap
+from pathlib import Path
+from typing import List
+from subprocess import run
+
+
+header_shim_sentinel = "// BOUT++ header shim"
+
+header_warning = f"""\
+#pragma once
+{header_shim_sentinel}
+#warning Header "{{0}}" has moved to "bout/{{0}}". Run `bin/bout-v5-header-upgrader.py` to fix
+#include "bout/{{0}}"
+"""
+
+
+def header_needs_moving(header: Path) -> bool:
+    """Check if `header` has not yet been moved"""
+    with open(header, "r") as f:
+        return header_shim_sentinel not in f.read()
+
+
+def deprecated_header_list(include_path: Path = Path("./include")):
+    """List of deprecated header paths (that is, those in bare
+    ``include`` directory)
+
+    """
+    return include_path.glob("*.hxx")
+
+
+def write_header_shim(header: Path):
+    """Write 'shim' for header, that ``include``s new location"""
+    with open(header, "w") as f:
+        f.write(header_warning.format(header.name))
+
+
+def fix_library_header_locations(
+    include_path: Path = Path("./include"), quiet: bool = False
+):
+    unmoved_headers = list(
+        filter(header_needs_moving, deprecated_header_list(include_path))
+    )
+    include_bout_path = include_path / "bout"
+
+    if unmoved_headers == []:
+        print("No headers to move!")
+        return
+
+    out = run("git diff-index --cached HEAD --quiet", shell=True)
+    if out.returncode:
+        raise RuntimeError(
+            "git index not clean! Please commit or discard any changes before continuing"
+        )
+
+    # First we move any existing headers and commit this change, so
+    # that history is preserved
+    for header in unmoved_headers:
+        new_path = include_bout_path / header.name
+        if not quiet:
+            print(f"Moving '{header}' to '{new_path}'")
+        run(f"git mv {header} {new_path}", shell=True, check=True)
+
+    run(r"git commit -m 'Move headers to `include/bout`'", shell=True, check=True)
+
+    # Now we can write the compatibility shim
+    for header in unmoved_headers:
+        write_header_shim(header)
+
+    run(f"git add {' '.join(map(str, unmoved_headers))}", shell=True, check=True)
+    run(
+        r"git commit -m 'Add compatibility shims for old header locations'",
+        shell=True,
+        check=True,
+    )
```

Comment:
Why not add as a single commit? That makes git-bisect less useful ...

Reply:
This is to try and make sure that git tracks the first commit as a rename. I found that doing this in a single commit (at least sometimes, in my limited testing) meant that git couldn't track changes across the commit.

You're right that it potentially makes bisects a bit more painful, but that's easier to work around with `git bisect skip`. Bisects are probably also (hopefully!) rarer than `git blame` for instance